### PR TITLE
9229 - Add bearer token auth mechanism for OIDC user with feature flag

### LINKED ISSUE #9229: As a developer, I'd like to use Bearer Token to send authenticated API requests
Placeholder for now. I'll extend it tomorrow. 

**Overview of the Feature Request**

**What kind of user is the feature intended for?**
(Example users roles: API User, Curator, Depositor, Guest, Superuser, Sysadmin)

**What inspired the request?**

**What existing behavior do you want changed?**

**Any brand new behavior do you want to add to Dataverse?**

**Any related open or closed issues to this feature request?**

### PR DESCRIPTION
**Pull request by johannes-darms**
- closes #9229
- replaces #9532

### CODE REVIEW

**Review by pdurbin — APPROVED** [review]
I tested the previous iteration of this pull request (#9532) and it worked so I'm approving this.

It looks like @johannes-darms incorporated all of my suggestions (thanks!).

@kcondon please note that the previous PR has all the details on how to test. I did my testing with Docker. I'm happy to walk your through what I did.

File: src/test/java/edu/harvard/iq/dataverse/api/auth/BearerTokenAuthMechanismTest.java
Comment (poikilotherm) [inline src/test/java/edu/harvard/iq/dataverse/api/auth/BearerTokenAuthMechanismTest.java:43]: Please move these `setUp()` to make the tests more readable. (Goes as well for all the other occurences...)

File: src/test/java/edu/harvard/iq/dataverse/api/auth/BearerTokenAuthMechanismTest.java
Comment (poikilotherm) [inline src/test/java/edu/harvard/iq/dataverse/api/auth/BearerTokenAuthMechanismTest.java:40]: I'm sorry that this is a bit messy - I did not yet implement setting something like this on a class level, this needs tuning. (Would avoid the repetition)

File: src/main/java/edu/harvard/iq/dataverse/api/auth/BearerTokenAuthMechanism.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/api/auth/BearerTokenAuthMechanism.java:58]: This is not covered by a test, but is part of the API contract. Please make sure there is a unit test for this.

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/oidc/OIDCAuthProvider.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/oidc/OIDCAuthProvider.java:283]: When are these exceptions thrown?

**Review by poikilotherm — CHANGES_REQUESTED** [review]
I'm sorry to be pedantic, but I do see some places that IMHO could result in a bad API user experience. Thanks for keeping up the good work.

File: src/main/java/edu/harvard/iq/dataverse/api/auth/BearerTokenAuthMechanism.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/api/auth/BearerTokenAuthMechanism.java:43]: @pdurbin @GPortas @kcondon @scolapasta should we address this now before merging or is it fine to let that go in as is? At least until now this is an experimental feature with a feature flag, so maybe OK to merge without this caching.

```suggestion
            // Validate and verify provided Bearer Token, and retrieve UserRecordIdentifier
            // TODO: Get the identifier from an invalidating cache to avoid lookup bursts of the same token. Tokens in the cache should be removed after some (configurable) time.
```

**Comment by pdurbin** [comment]
@poikilotherm I'm not sure if this was your intention by selecting "request changes" but merging is blocked until you switch to "comment" or "approve".

File: src/main/java/edu/harvard/iq/dataverse/api/auth/BearerTokenAuthMechanism.java
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/api/auth/BearerTokenAuthMechanism.java:55]:
```suggestion
                logger.log(Level.WARNING, "Bearer token detected, OIDC provider {0} validated Token but no linked UserAccount", userInfo.getUserRepoId());
                // TODO: Instead of returning null, we should throw a meaningful error to the client.
                // Probably this will be a wrapped auth error response with an error code and a string describing the problem.
```

@poikilotherm says you're willing to help with this. (We wrote this suggestion together.)

File: src/main/java/edu/harvard/iq/dataverse/api/auth/BearerTokenAuthMechanism.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/api/auth/BearerTokenAuthMechanism.java:92]: It sounds like the 401 is sent later because we aren't doing anything in the catch here.

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/oidc/OIDCAuthProvider.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/oidc/OIDCAuthProvider.java:285]: The idea is to just return an empty optional rather than throwing an OAuth2Exception.

**Review by pdurbin — COMMENTED** [review]
@poikilotherm and I chatted about this PR and left some comment.

**Pull request merged**
